File: II semestr/HomeWork4/Zadacha2 UniqueList/Zadacha2 UniqueList/List.cs
namespace UniqueListNamespace
{
    /// <summary>
    /// Class list.
    /// </summary>
    public class List
    {
        private class ListElement
        {
            /// <summary>
            /// Next element.
            /// </summary>
            public ListElement Next { get; set; }

            /// <summary>
            /// Value of list element.
            /// </summary>
            public int Value { get; set; }
        }

        private ListElement head;

        /// <summary>
        /// Insert value in the head of the list.
        /// </summary>
        /// <param name="value"></param>
        public virtual void InsertToHead(int value)
        {
            var newElement = new ListElement()
            {
                Value = value,
                Next = head
            };
            head = newElement;
        }

        /// <summary>
        /// Return value of the position.
        /// If position bigger than number of elements in list then return -1.
        /// </summary>
        /// <param name="position"></param>
        /// <returns></returns>
        public int ValueOfPosition(int position)
        {
            var element = head;
            if (element == null)
            {
                return -1;
            }
            for (int i = 0; i < position; i++)
            {
                if (element.Next == null)
                {
                    return -1;
                }
                element = element.Next;
            }
            return element.Value;
        }

        /// <summary>
        /// Insert value to position in list.
        /// If list is empty then insert to head.
        /// If number of elements in list less than posision then insert to end of list.
        /// </summary>
        /// <param name="value"></param>
        /// <param name="position"></param>
        public virtual void InsertToPosition(int value, int position)
        {
            var element = head;
            if (element == null || position == 0)
            {
                InsertToHead(value);
                return;
            }
            for (int i = 0; i < position - 1; i++)
            {
                if (element.Next == null)
                {
                    break;
                }
                element = element.Next;
            }
            var newElement = new ListElement()
            {
                Value = value,
                Next = element.Next
            };
            element.Next = newElement;
        }

        /// <summary>
        /// Remove first element with value from the list.
        /// </summary>
        /// <param name="value"></param>
        public virtual void RemoveElement(int value)
        {
            var element = head;
            if (element == null)
            {
                return;
            }
            if (element.Value == value)
            {
                head = head.Next;
                return;
            }
            while (element.Next != null)
            {
                if (element.Next.Value == value)
                {
                    element.Next = element.Next.Next;
                    return;
                }
                element = element.Next;
            }
        }

        /// <summary>
        /// Remove element to position in list.
        /// If list is empty then this will be written to the console.
        /// If position bigger then number of elements of the list then this will be written.
        /// </summary>
        /// <param name="position"></param>
        public void RemoveElementToPosition(int position)
        {
            if (position == 0)
            {
                head = head.Next;
                return;
            }
            var element = head;
            if (element == null)
            {
                return;
            }
            for (int i = 0; i < position - 1; i++)
            {
                if (element.Next == null)
                {
                    return;
                }
                element = element.Next;
            }
            element.Next = element.Next.Next;
        }

        /// <summary>
        /// Clean all list.
        /// </summary>
        public void RemoveList()
        {
            head = null;
        }

        /// <summary>
        /// Return 'true' if list is empty and 'false' if it is not empty.
        /// </summary>
        /// <returns></returns>
        public bool IsEmpty()
        {
            return (head == null);
        }

        /// <summary>
        /// Is value in the list?
        /// </summary>
        /// <param name="value"></param>
        /// <returns></returns>
        public bool Exist(int value)
        {
            var element = head;
            while (element != null)
            {
                if (element.Value == value)
                {
                    return true;
                }
                element = element.Next;
            }
            return false;
        }
    }
}

File: I semester/HomeWork7/Zadacha1Array/Zadacha1Array/list.cpp
#include <iostream>
#include "list.h"

using namespace std;

const int length = 100;

struct ListElement
{
	ElementType value[length];
	int number;
};

struct List
{
	ListElement *head;
};

List *create()
{
	List *result = new List;
	ListElement *element = new ListElement;
	element->number = 0;
	result->head = element;
	return result;
}

void insertToHead(List *list, ElementType value)
{
	list->head->value[list->head->number] = value;
	list->head->number++;
}

void insert(List *list, ElementType value, Position position)
{
	list->head->value[position] = value;
	list->head->number++;
}

void print(List *list)
{
	if (list->head->number == 0)
	{
		cout << "List is clear!\n";
		return;
	}
	cout << "List:\n";
	for (int i = 0; i < list->head->number; i++)
	{
		cout << list->head->value[i] << " ";
	}
	cout << "\n";
}

Position first(List *list)
{
	return 0;
}

Position next(List *list, Position position)
{
	return (position + 1);
}

ElementType returnValue(List *list, Position position)
{
	return list->head->value[position];
}

void deleteList(List *list)
{
	delete list->head;
	delete list;
}

bool noolPosition(List *list, Position position)
{
	return (position >= list->head->number);
}